From bd7fb17449accb9bf5cbc5d291ca5e3a1429be3d Mon Sep 17 00:00:00 2001
From: Jenny TC <jenny.tc@intel.com>
Date: Mon, 27 Oct 2014 09:13:38 +0530
Subject: [PATCH 1/1] PMIC: Add pdata for wcove GPADc.

Add pdata for whiskey cove GPADC driver which defines the regmap
and channel map for Wcove GPADC.

Change-Id: I2bbbe110f9d222d2f5db871a534e17e068799166
Signed-off-by: Jenny TC <jenny.tc@intel.com>
---
 drivers/mfd/intel_soc_pmic_wc.c |   95 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 95 insertions(+)

diff --git a/drivers/mfd/intel_soc_pmic_wc.c b/drivers/mfd/intel_soc_pmic_wc.c
index 519007ba072a..403a21d6896f 100644
--- a/drivers/mfd/intel_soc_pmic_wc.c
+++ b/drivers/mfd/intel_soc_pmic_wc.c
@@ -28,6 +28,10 @@
 #include <linux/acpi.h>
 #include <linux/version.h>
 #include <asm/intel_wcove_bcu.h>
+#include <linux/iio/intel_wcove_gpadc.h>
+#include <linux/iio/iio.h>
+#include <linux/iio/machine.h>
+#include <linux/iio/types.h>
 #include <linux/power/intel_pmic_ccsm.h>
 #include <linux/mfd/intel_soc_pmic_wcove.h>
 #include "./intel_soc_pmic_core.h"
@@ -55,6 +59,80 @@
 
 #define WCOVE_PMIC_I2C_ID 21
 
+static struct gpadc_regmap_t whiskeycove_gpadc_regmaps[GPADC_NUM_CHANNELS] = {
+	{"VBAT",	5,	0x4F03, 0x4F04, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"BATID",	4,	0x4F06, 0x4F07, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"PMICTEMP",	3,	0x4F42,	0x4F43, 0x4F33, 0x4F34, 0x4F33, 0x4F34},
+	{"BATTEMP0",	2,	0x4F15, 0x4F16, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"BATTEMP1",	2,	0x4F17, 0x4F18, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"SYSTEMP0",	3,	0x4F38, 0x4F39, 0x4F23, 0x4F24, 0x4F25, 0x4F26},
+	{"SYSTEMP1",	3,	0x4F3A, 0x4F3B, 0x4F27, 0x4F28, 0x4F29, 0x4F2A},
+	{"SYSTEMP2",	3,	0x4F3C, 0x4F3D, 0x4F2B, 0x4F2C, 0x4F2D, 0x4F2E},
+	{"USBID",	1,	0x4F08, 0x4F09, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"PEAK",	7,	0x4F13, 0x4F14, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"AGND",	6,	0x4F0A, 0x4F0B, 0xFF, 0xFF, 0xFF, 0xFF},
+	{"VREF",	6,	0x4F0A, 0x4F0B, 0xFF, 0xFF, 0xFF, 0xFF},
+};
+
+static struct gpadc_regs_t whiskeycove_gpadc_regs = {
+	.gpadcreq	=	0x4F02,
+	.gpadcreq_irqen	=	0,
+	.gpadcreq_busy	=	(1 << 0),
+	.mirqlvl1	=	0x6e0E,
+	.mirqlvl1_adc	=	(1 << 3),
+	.adc1cntl	=	0x4F05,
+	.adcirq		=	0x6E08,
+	.madcirq	=	0x6E15,
+};
+
+#define MSIC_ADC_MAP(_adc_channel_label,			\
+		     _consumer_dev_name,                        \
+		     _consumer_channel)                         \
+	{                                                       \
+		.adc_channel_label = _adc_channel_label,        \
+		.consumer_dev_name = _consumer_dev_name,        \
+		.consumer_channel = _consumer_channel,          \
+	}
+
+static struct iio_map wc_iio_maps[] = {
+	MSIC_ADC_MAP("CH0", "VIBAT", "VBAT"),
+	MSIC_ADC_MAP("CH1", "BATID", "BATID"),
+	MSIC_ADC_MAP("CH2", "PMICTEMP", "PMICTEMP"),
+	MSIC_ADC_MAP("CH3", "BATTEMP", "BATTEMP0"),
+	MSIC_ADC_MAP("CH4", "BATTEMP", "BATTEMP1"),
+	MSIC_ADC_MAP("CH5", "SYSTEMP", "SYSTEMP0"),
+	MSIC_ADC_MAP("CH6", "SYSTEMP", "SYSTEMP1"),
+	MSIC_ADC_MAP("CH7", "SYSTEMP", "SYSTEMP2"),
+	MSIC_ADC_MAP("CH8", "USBID", "USBID"),
+	MSIC_ADC_MAP("CH9", "PEAK", "PEAK"),
+	MSIC_ADC_MAP("CH10", "GPMEAS", "AGND"),
+	MSIC_ADC_MAP("CH11", "GPMEAS", "VREF"),
+	{ },
+};
+
+#define MSIC_ADC_CHANNEL(_type, _channel, _datasheet_name) \
+	{                               \
+		.indexed = 1,           \
+		.type = _type,          \
+		.channel = _channel,    \
+		.datasheet_name = _datasheet_name,      \
+	}
+
+static const struct iio_chan_spec const wc_adc_channels[] = {
+	MSIC_ADC_CHANNEL(IIO_VOLTAGE, 0, "CH0"),
+	MSIC_ADC_CHANNEL(IIO_VOLTAGE, 1, "CH1"),
+	MSIC_ADC_CHANNEL(IIO_TEMP, 2, "CH2"),
+	MSIC_ADC_CHANNEL(IIO_TEMP, 3, "CH3"),
+	MSIC_ADC_CHANNEL(IIO_TEMP, 4, "CH4"),
+	MSIC_ADC_CHANNEL(IIO_TEMP, 5, "CH5"),
+	MSIC_ADC_CHANNEL(IIO_TEMP, 6, "CH6"),
+	MSIC_ADC_CHANNEL(IIO_TEMP, 7, "CH7"),
+	MSIC_ADC_CHANNEL(IIO_VOLTAGE, 8, "CH8"),
+	MSIC_ADC_CHANNEL(IIO_VOLTAGE, 9, "CH9"),
+	MSIC_ADC_CHANNEL(IIO_VOLTAGE, 10, "CH10"),
+	MSIC_ADC_CHANNEL(IIO_VOLTAGE, 11, "CH11"),
+};
+
 enum {
 	PWRSRC_LVL1 = 0,
 	THRM_LVL1,
@@ -75,6 +153,7 @@ enum {
 };
 
 struct intel_soc_pmic whiskey_cove_pmic;
+
 static struct temp_lookup th05_lookup_tbl[] = {
 	{2241, 125, 0}, {2541, 120, 0},
 	{2893, 115, 0}, {3307, 110, 0},
@@ -427,6 +506,21 @@ struct intel_pmic_irqregmap whiskey_cove_irqregmap[] = {
 	},
 };
 
+static void wc_set_adc_pdata(void)
+{
+	static struct intel_wcove_gpadc_platform_data wc_adc_pdata;
+	wc_adc_pdata.channel_num = GPADC_NUM_CHANNELS;
+	wc_adc_pdata.intr_mask = MUSBID | MPEAK | MBATTEMP
+		| MSYSTEMP | MBATT | MVIBATT | MGPMEAS | MCCTICK;
+	wc_adc_pdata.gpadc_iio_maps = wc_iio_maps;
+	wc_adc_pdata.gpadc_regmaps = whiskeycove_gpadc_regmaps;
+	wc_adc_pdata.gpadc_regs = &whiskeycove_gpadc_regs;
+	wc_adc_pdata.gpadc_channels = wc_adc_channels;
+
+	intel_soc_pmic_set_pdata("wcove_gpadc", (void *)&wc_adc_pdata,
+			sizeof(wc_adc_pdata), 0);
+}
+
 static void wcove_set_ccsm_config(void)
 {
 	static struct intel_pmic_ccsm_platform_data pdata;
@@ -453,6 +547,7 @@ static int whiskey_cove_init(void)
 
 	wcove_set_ccsm_config();
 	wcove_set_bcu_pdata();
+	wc_set_adc_pdata();
 
 	return 0;
 }
-- 
1.7.9.5

